Add micrometer support for Oracle Cloud Monitoring

pgressa opened the pull request.

pgressa commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactory.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    @Singleton
+    OracleCloudMeterRegistry oracleCloudMeterRegistry(ExportConfigurationProperties exportConfigurationProperties) {
+        String ingestionEndpoint = String.format("https://telemetry-ingestion.%s.oraclecloud.com", regionProvider.getRegion().getRegionId());
```
For posting metrics, we need to use different endpoint. This is just a working version for testing, I didn't come up with any trick how to do it nicely. If we would stay with this solution, I will have to inject also other configuration related beans to for the client builder.

graemerocher replied:
I don't understand, can't you just make the endpoint configurable via a configurable property or class?

pgressa replied:
There are several issues.

First is that the SDK factory generates singletons, so in case if the monitoring client would be directly configured to use ingestion endpoint to post metrics, the other consumers of monitoring client wouldn't have to use  "get" request (I didn't find direct note on that the get requests wouldn't work, but if there are two distinct endpoints and one is specifically for ingestion, then I wouldn't rely on that such endpoint will always processes get requests). The ingestion endpoint is specifically for one specific method [postMetricData](https://docs.oracle.com/en-us/iaas/tools/java/1.32.1/com/oracle/bmc/monitoring/MonitoringClient.html#postMetricData-com.oracle.bmc.monitoring.requests.PostMetricDataRequest).

Secondly there are some ways how to influence the final rest client configuration trough [ClientConfigurator](https://docs.oracle.com/en-us/iaas/tools/java/1.32.1/com/oracle/bmc/http/ClientConfigurator.html) or [ClientConfiguration](https://docs.oracle.com/en-us/iaas/tools/java/1.32.1/com/oracle/bmc/ClientConfiguration.html), but none of them allows me to change the endpoint- plus again we would have to distinguish somehow that given configurator is for telemetry-ingestion. 

From the above either we will create specific MonitoringIngestionClient class, so it would be injectable and would have everything configuration, or we will build ingestion client in the factory like now. Of course I'm all ears for some other tricks, but at this moment I can't imagine how to distinguish in between ingestion and regular client on configuration or bean level.

graemerocher replied:
Building a specific `MonitoringIngestionClient` seems to make sense to me. You could always allow this bean to be replaced for customisability

pgressa commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactory.java:
```diff
@@ -0,0 +1,88 @@
+                endpoint(ingestionEndpoint).
+                build(authenticationDetailsProvider);
+
+        exportConfigurationProperties.getExport().computeIfAbsent("oraclecloudmonitoring.compartmentId", x -> {
```
CompartmentId is required in order to push metrics, this is kind of default approach that will be mentioned in docs, that by default compartmentId from auth configuration is taken.

pgressa commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/micrometer/MetricDataDetailsPartition.java:
```diff
@@ -0,0 +1,39 @@
+        super(list, partitionSize);
+    }
+
+    static List<List<MetricDataDetails>> partition(List<MetricDataDetails> list, int partitionSize) {
```
using this as there's limit of 50 metrics send in one post metric details request.

pgressa commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/micrometer/OracleCloudMeterRegistry.java:
```diff
@@ -0,0 +1,291 @@
+        this.oracleCloudConfig = oracleCloudConfig;
+
+        config().namingConvention(new OracleCloudMetricsNamingConvention());
+        config().commonTags("application", "micronaut");
```
The metric tags are transformed into oracle cloud metric dimensions, and there is at least one required dimension for every metric data details. I think this configuration should be externalised into exporter config, but I'm not sure about the tag key.

graemerocher replied:
Maybe this should be looked up from `ApplicationConfiguration` where you can get the application name?

pgressa replied:
👍

graemerocher commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/micrometer/MetricDataDetailsPartition.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Pavol Gressa
+ * @since 2.3
+ */
+public class MetricDataDetailsPartition extends AbstractPartition<MetricDataDetails> {
```
Make as `@Internal`. Does this need to be public?

graemerocher commented on oraclecloud-monitoring/src/main/java/io/micronaut/oraclecloud/monitoring/micrometer/OracleCloudConfig.java:
```diff
@@ -0,0 +1,77 @@
+        return "oraclecloudmonitoring";
+    }
+
+    default String batchAtomicity() {
+        return getString(this, "batchAtomicity").orElse(null);
+    }
+
+    default String namespace() {
+        return getString(this, "namespace").required().get();
+    }
+
+    default String compartmentId() {
+        return getString(this, "compartmentId").required().get();
+    }
+
+    default String resourceGroup() {
+        return getString(this, "resourceGroup").orElse(null);
+    }
+
+    default int batchSize() {
+        return getInteger(this, "batchSize").orElse(50);
+    }
+
```
Missing javadoc for these methods

graemerocher reviewed (changes requested):
Looks good, just a few tweaks and documentation is needed.

recursivecodes reviewed (approved):
Looks good to me.

pgressa commented:
@recursivecodes @graemerocher guys, check pls https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/126/commits/c82f58b6dbf221b7110e4b2248c5a6d08349019c .. added docs, renamed module and there's MonitoringIngestionClient. I would say it's not terrible, not great solution how to solve separation of concerns of MonitoringClient. This solution should allow you to inherit the class and provide your own implementation how to create monitoring client. Tried few approaches, this one, at the end, seems most simple and easy to read. If you like I can create interface, then extend the interface  by monitoring client, and create also factory for the ingestion monitoring client. But I wonder whether it's worth it.
